_blog/2024-02-22-api-rate-limiting-rust.mdx

```diff
@@ -0,0 +1,265 @@
+date: '2024-02-22T15:00:00'
+---
+
+Hello world! We’re going to talk about implementing rate limiting for your API in Rust. When it comes to services in production, you want to ensure that bad actors aren’t abusing your APIs - this is where API rate limiting comes in.
+
+For this tutorial we will be implementing the “sliding window” algorithm by having a dynamic period to check request histories over as well, using a basic in-memory hashmap to store users and their request times. We will also look at using `tower-governor` to configure rate limiting for you.
+
+## Getting started
+
+To get started, we’re going to initialise a Shuttle project using `cargo shuttle init` and follow the prompt, picking Axum as our framework of choice. 
+
+We’re going to need some extra dependencies, so let’s install them with this shell snippet:
+
+```bash
+cargo add serde@1.0.196 -F derive
+cargo add chrono@0.4.34 -F serde,clock
+```
+
+Once done, we can get started!
+
+## Implementing a naive sliding window rate limiter
+
+To figure out how we can do this from the ground up, let’s write a naive sliding window IP based rate limiter from scratch.
+
+We’ll declare a new struct that holds a `HashMap` of `IpAddr` keys with the values being `Vec<DateTime<Utc>>` (a Vector of UTC-timezone timestamps). 
+
+```rust
+use std::sync::{Arc, Mutex};
+use std::collections::HashMap;
+use std::net::IpAddr;
+use chrono::{DateTime, Utc};
+
+// This will be the request limit (per minute) for a user to access an endpoint
+// If the user attempts to go beyond this limit, we should return an error
+const REQUEST_LIMIT: usize = 120;
+
+#[derive(Clone, Default)]
+pub struct RateLimiter {
+    requests: Arc<Mutex<HashMap<IpAddr, Vec<DateTime<Utc>>>>>,
+}
+```
+
+ To get started, we’ll want to lock our hashmap by using `.lock()` which gives us write access. Then we’ll want to check whether or not the hashmap contains a key containing the IP address we want to check for with the `.entry()` function, then modify it by retaining valid timestamps and pushing a new entry depending on whether or not the length is under the request limit. We then check if the entry length is a higher length than the request limit - if so, return an error; if not, return `Ok(())`.
+
+```rust
+impl RateLimiter {
+    fn check_if_rate_limited(&self, ip_addr: IpAddr) -> Result<(), String> {
+        // we only want timestamps from up to 60 seconds ago
+        let throttle_time_limit = Utc::now() - std::time::Duration::from_secs(60);
+
+        let mut requests_hashmap = self.requests.lock().unwrap();
+
+        let requests_for_ip = requests_hashmap
+            // grab the entry here and allow us to modify it in place
+            .entry(ip_addr)
+            // modify the entry in place here
+            .and_modify(|x| {
+                if x.len() < REQUEST_LIMIT {
+                    x.retain(|x| x.to_utc() > throttle_time_limit);
+                    x.push(Utc::now());
+                }
```

Comment:
Should we not retain regardless of the vector length? This if-case makes the vec grow untill "full" and then nothing happens (always rate limited).